driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -23,64 +23,68 @@
 import com.mongodb.ServerCursor;
 import com.mongodb.binding.ReadBinding;
 import org.bson.BsonDocument;
+import org.bson.BsonTimestamp;
 import org.bson.RawBsonDocument;
 
 import java.util.ArrayList;
 import java.util.List;
 
 import static com.mongodb.operation.ChangeStreamBatchCursorHelper.isRetryableError;
 
-final class ChangeStreamBatchCursor<T> implements BatchCursor<T> {
+final class ChangeStreamBatchCursor<T> implements ChangeStreamCursor<T> {
     private final ReadBinding binding;
     private final ChangeStreamOperation<T> changeStreamOperation;
 
-    private BsonDocument resumeToken;
-    private BatchCursor<RawBsonDocument> wrapped;
+    private BsonTimestamp startAtOperationTime;
+    private ChangeStreamCursor<RawBsonDocument> wrapped;
+    private BsonDocument postBatchResumeToken;
 
     ChangeStreamBatchCursor(final ChangeStreamOperation<T> changeStreamOperation,
-                            final BatchCursor<RawBsonDocument> wrapped,
+                            final ChangeStreamCursor<RawBsonDocument> wrapped,
                             final ReadBinding binding) {
         this.changeStreamOperation = changeStreamOperation;
-        this.resumeToken = changeStreamOperation.getResumeToken();
+        this.startAtOperationTime = changeStreamOperation.getStartAtOperationTime();
         if (changeStreamOperation.getStartAfter() != null) {
-            this.resumeToken = changeStreamOperation.getStartAfter();
+            changeStreamOperation.resumeToken(changeStreamOperation.getStartAfter());
```

Comment:
Possibly worth nulling `startAfter` here - then you dont have to worry about it again.

Reply:
Done